parsec/mca/device/cuda/device_cuda_component.c

```diff
@@ -141,7 +141,7 @@ static int device_cuda_component_query(mca_base_module_t **module, int *priority
                 PARSEC_CUDA_CHECK_ERROR( "(parsec_device_cuda_component_query) cuCtxEnablePeerAccess ", cudastatus,
                                          {continue;} );
                 source_gpu->super.peer_access_mask = (int16_t)(source_gpu->super.peer_access_mask | (int16_t)(1 <<
-                        target_gpu->cuda_index));
+                        target_gpu->super.super.device_index));
```

Comment:
This is a major change ! You repurpose the device index bitmask from a device naming scheme (`cuda_index` being the CUDA naming of the device) to a parsec naming scheme (`device_index` being the position in the parsec's array of devices).

However, I don't see any changes on the use of this `peer_access_mask` which point to something weird in this patch.

Reply:
I did change the initialization of `peer_access_mask` in `device_cuda_component.c:143`: we use `target_gpu->super.super.device_index` instead of `target_gpu->cuda_index`, so I believe the access and initialization are consistent.

In parsec master, `peer_access_mask` is 
 - initialized to 0 in `parsec_cuda_module_init`  in `device_cuda_module.c`
 - then computed in `device_cuda_component_query` in `device_cuda_component.c` after all modules have been loaded
 - then read in `parsec_cuda_data_stage_in` in `device_cuda_module.c`

In this PR, `parsec_cuda_data_stage_in` is moved in `device_gpu.c`, and I have changed for both the access and initialization what we use as index.



The user-facing MCA parameter  `device_cuda_nvlink_mask` is still using the CUDA index to exclude pairs of devices that could rely on NVLINK (so things have not changed here), and we only display `peer_access_mask` in debugging functions (the mask show parsec-device indices instead of cuda-indices, so it's not a 1-to-1 with the MCA parameter).



If this is not acceptable, to make `parsec_gpu_data_stage_in` device-type independent and keep it in `device_gpu.c`, I can extend the API of the devices to provide a function `bool device_to_device_direct(parsec_device_module_t *src, parsec_device_module_t *dst)` and each device will fill it up. But since `src` and `dst` might be of different types, the first test will be to return `false` if `src->type != dst->type`.